### PR TITLE
Optimizations for dealing with large datasets

### PR DESCRIPTION
**Pull request by dmerrill6**
# Change log

- Reduced the amount of database roundtrips required to build up the complete graphql response (now 2 queries are required if fetching `totalCount` and 1 otherwise). This was done by skipping calls to `getNodesLength` and calculating node operations with less info.

- Added an option to skip fetching `totalCount` for large databases.

- Added `startCursor` and `endCursor` to the response payload.

image

### CODE REVIEW

File: tests/test-app/compiled-dist/models/db.js [hunk 1]
```diff
@@ -5,5 +5,7 @@ const Knex = require('knex');
 const knexOptions = require('../../knexfile');
 
 const env = process.env.NODE_ENV || 'development';
-const knex = Knex(knexOptions[env]);
+const knex = Knex(knexOptions[env]); // for debugging
+
+knex.on('query', console.log);
```
Comment (dmolina79) [inline tests/test-app/compiled-dist/models/db.js]: does this need to be flagged?

File: tests/test-app/tests/apollo-cursor-pagination/knex-offset-implementation.test.js [hunk 2]
```diff
@@ -0,0 +1,361 @@
+const _ = require('../helpers/database'); // eslint-disable-line no-unused-vars
+const { start, stop, graphqlQuery } = require('../helpers/integration-server');
+const catFactory = require('../factories/mocks/cat');
+
+describe('getCatsByOwner root query', () => {
```
Comment (dmolina79) [inline tests/test-app/tests/apollo-cursor-pagination/knex-offset-implementation.test.js]: do you have the image of the tests passing?

Comment (dmerrill6) [reply tests/test-app/tests/apollo-cursor-pagination/knex-offset-implementation.test.js]: Added in the description

File: tests/test-app/compiled-dist/models/db.js (hunk 1 above)
Comment (dmerrill6) [reply tests/test-app/compiled-dist/models/db.js]: Removed in e7d44d5

**Pull request merged**